rm/rm-server/src/test/java/org/ow2/proactive/resourcemanager/frontend/topology/TestTopologyConcurrency.java

```diff
@@ -0,0 +1,186 @@
+import org.ow2.proactive.resourcemanager.core.properties.PAResourceManagerProperties;
+import org.ow2.proactive.resourcemanager.frontend.topology.pinging.HostsPinger;
+import org.ow2.proactive.resourcemanager.rmnode.RMNode;
+import org.ow2.proactive.resourcemanager.selection.SelectionManager;
+import org.ow2.proactive.resourcemanager.selection.topology.TopologyManager;
+import org.ow2.proactive.scripting.ScriptResult;
+import org.ow2.proactive.scripting.SelectionScript;
+
+import java.io.IOException;
+import java.lang.reflect.InvocationTargetException;
+import java.net.InetAddress;
+import java.rmi.AlreadyBoundException;
+import java.rmi.dgc.VMID;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+
+/**
+ * Test add and remove nodes from multiple threads
+ * <p/>
+ * At the end, remove all previously added nodes and verify that
+ * the topology manager structure is empty
+ */
+public class TestTopologyConcurrency {
+
+    // number of thread used in this test
+    final int nb_threads = 20;
+
+    // multiplier for the number of nodes to create
+    final int node_factor = 4;
+
+    // size of collisions, as many operations will be done on the same node
+    // more removal than add
+    final int collision_size = 10;
+
+    // total number of tasks
+    final int total = nb_threads * node_factor * collision_size;
+
+    String baseUrl = "pnp://localhost:1234/LocalNode-";
+
+    ExecutorService s = Executors.newFixedThreadPool(nb_threads);
+
+    @Test
+    public void action() throws Exception {
+
+        BasicConfigurator.resetConfiguration();
+        BasicConfigurator.configure();
+        Logger.getLogger(TopologyManager.class).setLevel(Level.DEBUG);
+        PAResourceManagerProperties.RM_TOPOLOGY_PINGER.updateProperty(HostsPinger.class.getName());
+        PAResourceManagerProperties.RM_TOPOLOGY_ENABLED.updateProperty("true");
+        PAResourceManagerProperties.RM_TOPOLOGY_DISTANCE_ENABLED.updateProperty("false");
+
+        final ProActiveRuntimeImpl runtime = mock(ProActiveRuntimeImpl.class);
+        when(runtime.getVMInformation()).thenReturn(new DummyVMInfo());
+        final TopologyManager manager = new TopologyManager();
+        List<Callable<Boolean>> calls = new ArrayList<>(total);
+
+        // first set of tasks, add nodes and remove them
+        for (int i = 0; i < total; i++) {
+            final int j = i;
+            calls.add(new Callable<Boolean>() {
+                @Override
+                public Boolean call() throws Exception {
+                    int index = j - (j % collision_size);
+                    if ((j % (collision_size / 2)) == 0) {
+                        manager.addNode(new NodeImpl(runtime, baseUrl + index));
```

Comment:
Could extract `new NodeImpl(runtime, baseUrl + index)` in a variable to avoid code duplication.